pandas/core/frame.py

```diff
@@ -6243,11 +6244,10 @@ def diff(self, periods=1, axis=0):
     # Function application
 
     def _gotitem(self,
-                 key,           # type: Union[str, List[str]]
-                 ndim,          # type: int
-                 subset=None    # type: Union[Series, DataFrame, None]
-                 ):
-        # type: (...) -> Union[Series, DataFrame]
+                 key: Union[str, List[str]],
+                 ndim: int,
+                 subset: Union[Series, 'DataFrame', None] = None,
```

Comment:
Makes sense to do this here though I assume we'd also want to couple this with the TYPE_CHECKING constant:

https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING

Though that wasn't introduced until 3.5.2. @jreback does our min support for 3.5 only cover latest bug fix release of that version or would we need to straddle support for 3.5.0 and 3.5.1?

Reply:
This case is similar to the `Index` change request above. `DataFrame` won't be defined until the whole class is executed, so we need a forward reference.

I'm not sure what we would use the TYPE_CHECKING constant for here. I got the impression that it's often used to support forward references to external modules rather than the module being checked. In this case, e.g., we wouldn't import frame into itself.

Reply:
According to the docs its used for 3rd party type checkers which ultimately is our goal here. 
Maybe OK for now I suppose could always revert to comment format is required for 3.5.0 and 3.5.1 support when we do set up static analysis

Reply:
@gwrome whenever you fix the `def type` type comment can you also make this `Optional` instead of a `Union` with `None`? Should be more idiomatic that way.

Reply:
Done and done